src/builder.rs

```diff
@@ -0,0 +1,472 @@
+            );
+
+            self.spends
+                .into_iter()
+                .zip(self.recipients.into_iter())
+                .map(|(spend, recipient)| ActionInfo::new(spend, recipient, &mut rng))
+                .collect()
+        };
+
+        // Move some things out of self that we will need.
+        let flags = self.flags;
+        let anchor = self.anchor;
+
+        // Determine the value balance for this bundle, ensuring it is valid.
+        let value_balance: ValueSum = pre_actions
+            .iter()
+            .fold(Ok(ValueSum::default()), |acc, action| {
+                acc? + action.value_sum()?
+            })?;
+
+        // Compute the transaction binding signing key.
+        let bsk = pre_actions
+            .iter()
+            .map(|a| &a.rcv)
+            .sum::<ValueCommitTrapdoor>()
+            .into_bsk();
+
+        // Create the actions.
+        let (actions, circuits): (Vec<_>, Vec<_>) =
+            pre_actions.into_iter().map(|a| a.build(&mut rng)).unzip();
+
+        // Verify that bsk and bvk are consistent.
+        let bvk = (actions.iter().map(|a| a.cv_net()).sum::<ValueCommitment>()
+            - ValueCommitment::derive(value_balance, ValueCommitTrapdoor::zero()))
+        .into_bvk();
+        assert_eq!(redpallas::VerificationKey::from(&bsk), bvk);
+
+        // Create the proof.
+        let instances: Vec<_> = actions
+            .iter()
+            .map(|a| a.to_instance(flags, anchor.clone()))
+            .collect();
+        let proof = Proof::create(pk, &circuits, &instances)?;
+
+        Ok(Bundle::from_parts(
+            NonEmpty::from_vec(actions).unwrap(),
+            flags,
+            value_balance,
+            anchor,
+            Unauthorized { proof, bsk },
+        ))
+    }
+}
+
+/// Marker for an unauthorized bundle, with a proof but no signatures.
+#[derive(Debug)]
+pub struct Unauthorized {
+    proof: Proof,
+    bsk: redpallas::SigningKey<Binding>,
+}
+
+impl Authorization for Unauthorized {
+    type SpendAuth = (Option<SpendAuthorizingKey>, SpendValidatingKey);
+}
+
+/// Marker for a partially-authorized bundle, in the process of being signed.
+#[derive(Debug)]
+pub struct PartiallyAuthorized {
+    proof: Proof,
+    binding_signature: redpallas::Signature<Binding>,
+    sighash: [u8; 32],
+}
+
+impl Authorization for PartiallyAuthorized {
+    type SpendAuth = (Option<redpallas::Signature<SpendAuth>>, SpendValidatingKey);
+}
+
+impl Bundle<Unauthorized> {
+    /// Loads the sighash into this bundle, preparing it for signing.
+    ///
+    /// This API ensures that all signatures are created over the same sighash.
+    pub fn prepare<R: rand_7::RngCore + rand_7::CryptoRng>(
+        self,
+        mut rng: R,
+        sighash: [u8; 32],
+    ) -> Bundle<PartiallyAuthorized> {
+        self.authorize(
+            &mut rng,
+            |rng, _, (dummy_ask, ak)| {
+                (
+                    if let Some(ask) = dummy_ask {
```

Comment:
Is there a good reason to do part of the work (the signatures for the dummy spends) now, and then part of it later? Do we not have `ask` later?

Reply:
We need to do it in an automated fashion (as the exact number of dummies is opaque to the caller), so it's either going to be triggered by `prepare` or `finalize`. I didn't see any point in holding onto the dummy `ask` values beyond `prepare`, and signing at that point slots the signatures nicely into the spots where `sign` is about to place the real ones.